### ISSUE TITLE
Chore: Make sure every fixable rule has output property in invalid tests

### ISSUE THREAD

**Issue by gyandeeps** [eslint/eslint#5962]
Since we have meta in most of rules now, I think its easy to check if the rule is fixable or not.
I want to enhance our rule tester to check if all invalid tests have `output` property or not. A lot of times I had to check manually whether contributors checked for auto fix code or not.

Its should fail that test which has missing output with error message saying output missing or something.

**Comment by platinumazure** [comment]
I'm :+1: in general on this, but in theory it's possible for rules to only offer fixes some of the time. Would it be better to require either output or some kind of "nothing should change/no fix should be offered" property?

**Comment by gyandeeps** [comment]
If a rule is fixable then its fixable all the way with no if's. So any rule which is fixable, we safely check if all the invalid cases have `output` property or not.

**Comment by platinumazure** [comment]
@gyandeeps That's not true. Any `context.report()` call may or may not have a `fix` property, so it is absolutely possible for a rule not to supply auto-fixes in all cases if there are multiple calls to `context.report()`.

Also, rules could have code like this:

```js
function report(node) {
    context.report({
        node: node,
        message: MESSAGE,
        fix: getFix(node)
    });
}

function getFix(node) {
    if (cannotFix()) {
        return null;
    }
    // Return a fixer function
}
```

**Comment by gyandeeps** [comment]
Its technically possible, but we as a team do not allow that as of today (AFAIK). If a rule is fixable then its fixable all the way. Can you point me to a rule which does partial fixing?

**Comment by platinumazure** [comment]
@gyandeeps You are correct that this doesn't happen in the eslint repo. However, this change could be a breaking change for plugin authors if they use RuleTester and dev-depend on latest ESLint, and neither of us can guarantee that all plugin rules follow the same convention as the ESLint repository. (This change is only breaking if they use the new metadata, so it won't have too widespread impact, but it's worth calling out.)

Like I said originally, I like the idea-- I just want to watch out for our users, who may find themselves unduly burdened in cases where a rule does not always issue a fix despite being labeled fixable.

**Comment by platinumazure** [comment]
Oh, I just thought of something. Could we make this a setting which is overrideable in RuleTester's defaultConfig option? If so, we could make this opt-in and ESLint could use it but plugin authors would be able to control it.

**Comment by alberto** [comment]
It it is enforced as part of our build in `npm test` that would not be a problem, and would allow us to run tests for a single file without failing (I usually don't add all fixes at once).

Other than that 👍

**Comment by nzakas** [comment]
One caveat: a rule might not fix everything it finds to be incorrect. For instance, we might only replace `let` with `const` when it is initialized. In that case, should the test still have an `output`? (This is actually why I didn't enforce this from the start.)

**Comment by nzakas** [comment]
Sorry, hit return too soon. :) We cannot assume that all rules ever written will always fix every problem they find.

And remember, the RuleTester is used by more than just us, so any change must make sense universally.

**Comment by gyandeeps** [comment]
I think we can skip this for now.
thanks everyone.
